File: SelectionSort/selection_sort.py
# @author：noc @time：2019年12月19日
# 选择排序

def findSmallest(arr):
    '''
    :param arr: 数组
    :return: 最小值下标
    '''
    smallest = arr[0]
    smallest_index = 0
    for i in range(1, len(arr)):
        if arr[i] < smallest:
            smallest = arr[i]
            smallest_index = i
    return  smallest_index

def solution(arr):
    '''
    :param arr: 数组
    :return: 排序后的数组
    '''
    newArr = []
    for i in range(len(arr)):
        smallest = findSmallest(arr)
        newArr.append(arr.pop(smallest))   # 删除arr的最小值并将该值加入到新数组
    return newArr

if __name__ == '__main__':
    print(solution([2,5,1,6,3,9]))

File: SelectionSort/selection_sort.java
import java.util.Arrays;
public class selection_sort {
    public static void main(String[] args) {
        int[] arr = {4,2,6,8,3,9};
        arr = solution(arr, arr.length);
        System.out.println(Arrays.toString(arr));
    }

    // 找到数组的最小值，并返回下标
    public static int findSmallest(int[] arr) {
        int smallest = arr[0];
        int smallest_index = 0;
        for (int i = 1; i < arr.length; i++){
            if (arr[i] < smallest){
                smallest = arr[i];
                smallest_index = i;
            }
        }
        return smallest_index;
    }

    // 删除最小值
    static int[] del(int[] arr, int e){
        arr[e] = arr[arr.length - 1];     // 将最后一个值赋值给要删除
        arr = Arrays.copyOf(arr, arr.length - 1);
        return arr;
    }

    public static int[] solution(int[] arr, int arrLength) {
        int[] newArr = new int[arrLength];
        int smallest;
        int n = arr.length;
        for (int i = 0; i < n; i++){
            smallest = findSmallest(arr);
            newArr[i] = arr[smallest];

            arr = del(arr, smallest);
        }
        return newArr;
    }
}
